### PR TITLE
Adding validation for delegates field. See issue #240

### PR DESCRIPTION
**Pull request by pedrosino**
Also check #243 for future polishing...

### CODE REVIEW

File: WcaOnRails/app/models/competition.rb [hunk 1]
```diff
@@ -43,6 +43,12 @@ def must_have_at_least_one_event
       errors.add(:eventSpecs, "Competition must have at least one event")
     end
   end
+  validate :must_have_at_least_one_delegate, if: :isConfirmed?
+  def must_have_at_least_one_delegate
+    if delegate_ids.length == 0
+      errors.add(:delegates, "Competition must have at least one WCA delegate")
```
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: It's weird that you check `delegate_ids`, but then add an error to the `delegates` field. If you change this to `delegate_ids` in both places, do you get a nice red box when there's an error?

Comment (pedrosino) [inline WcaOnRails/app/models/competition.rb]: Indeed I do.
Why is there a delegates field? lol

I tried doing the same for events, but there is no f.input for them, just a label.

Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: Yeah, events are special, it should be pretty easy to write some code that checks `competition.errors.has_key? "event_specs"`, and shows the error message, though.

Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: The delegates method is a magic method from Rails that looks up all the delegates in the `competition_delegates` table and returns the delegates as an array. See https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/app/models/competition.rb#L11 for the line of code that tells ActiveRecord to create this delegates method.

File: WcaOnRails/app/assets/stylesheets/competitions.scss [hunk 2]
```diff
@@ -69,3 +69,8 @@ $venue-map-wrapper-height: 400px;
     display: inline;
   }
 }
+
+.modal-wide {
+  overflow-y: auto;
+  width: 600px;
```
Comment (jfly) [inline WcaOnRails/app/assets/stylesheets/competitions.scss]: This seems bad for mobile. How does this look on a phone?

File: WcaOnRails/app/models/competition.rb [hunk 3]
```diff
@@ -40,7 +40,13 @@ class Competition < ActiveRecord::Base
   validate :must_have_at_least_one_event, if: :isConfirmed?
   def must_have_at_least_one_event
     if events.length == 0
-      errors.add(:eventSpecs, "Competition must have at least one event")
+      errors.add(:Events, "Competition must have at least one event")
```
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: What's `Events` with a capital E?

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb [hunk 4]
```diff
@@ -54,8 +54,35 @@
       Last <%= Competition::NEARBY_DAYS_INFO %> days
     </label>
     <div class="col-sm-9">
-      <% nearby_count = @competition.nearby_competitions(Competition::NEARBY_DAYS_INFO, Competition::NEARBY_DISTANCE_KM_DANGER).count %>
-      <div class="<%= nearby_count < Competition::NEARBY_INFO_COUNT ? "alert alert-success" : "alert alert-danger" %>" role="alert"><%= nearby_count %> <%= "competition".pluralize(nearby_count) %> within <%= Competition::NEARBY_DISTANCE_KM_DANGER %>km and <%= Competition::NEARBY_DAYS_INFO %> days.
+      <% nearby_info = @competition.nearby_competitions(Competition::NEARBY_DAYS_INFO, Competition::NEARBY_DISTANCE_KM_DANGER).sort_by { |i| i.start_date } %>
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Does `...sort_by(&:start_date)` work here instead?

I don't like that we're using `NEARBY_DAYS_INFO` alongside `NEARBY_DISTANCE_KM_DANGER`. Can we use `NEARBY_DAYS_INFO` and `NEARBY_DISTANCE_KM_INFO` instead?

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb [hunk 5]
```diff
@@ -54,8 +54,35 @@
...
+      <% nearby_count = nearby_info.count %>
+      <div class="<%= nearby_count < Competition::NEARBY_INFO_COUNT ? "alert alert-success" : "alert alert-danger" %>" role="alert" style="cursor: pointer;" data-toggle="modal" data-target=".bs-example-modal-sm"><%= nearby_count %> <%= "competition".pluralize(nearby_count) %> within <%= Competition::NEARBY_DISTANCE_KM_DANGER %>km and <%= Competition::NEARBY_DAYS_INFO %> days.
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: I see `"alert ` in 2 places here, how about pulling that out to just after the `class="`?

I'm not okay with the inline css style. We don't do that here =).

As before, let's change `NEARBY_DISTANCE_KM_DANGER` to `NEARBY_DISTANCE_KM_INFO` here

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb [hunk 6]
```diff
@@ -54,8 +54,35 @@
...
+      </div>
+    </div>
+  </div>
+
+  <div id="" class="modal fade bs-example-modal-sm" tabindex="-1" role="dialog" aria-labelledby="mySmallModalLabel" aria-hidden="true">
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: What is `mySmallModalLabel`?

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb [hunk 7]
```diff
@@ -54,8 +54,35 @@
...
+    <div class="modal-dialog modal-sm modal-wide">
+      <div class="modal-content">
+        <div class="modal-header">
+          <h4 class="modal-title">Within 365 days</h4>
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Ah! What is this number 365? Let's change it to a variable.

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb [hunk 8]
```diff
@@ -54,8 +54,35 @@
...
+	          <th>Name</th>
+	          <th>Date</th>
+	          <th>Distance</th>
+	        </tr>
+	        <% nearby_info.each do |i| %>
+	          <tr>
+	            <td><%= i.id %></td>
+	            <td>
+	              <% days_until = (i.start_date - @competition.start_date).to_i %>
+	              <%= days_until.abs %> <%="day".pluralize(days_until) %> <%= days_until < 0 ? "before" : "after" %></td>
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Nitpick: there's no space between the equals sign and the opening quote in "day"

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb [hunk 9]
```diff
@@ -54,8 +54,35 @@
...
+        <div class="modal-header">
+          <h4 class="modal-title">Within 365 days</h4>
+        </div>
+	      <table class="table">
+	        <tr>
+	          <th>Name</th>
+	          <th>Date</th>
+	          <th>Distance</th>
+	        </tr>
+	        <% nearby_info.each do |i| %>
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Can we give `i` a better name than this? It's a nearby competition, right? How about `nearby_competition`?

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb [hunk 10]
```diff
@@ -64,6 +91,7 @@
 <script>
   (function() {
     $('[data-toggle="tooltip"]').tooltip();
+    $('#myModal').modal('show');
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: What is `myModal`?

File: WcaOnRails/app/models/competition.rb (hunk 3 above)
Comment (pedrosino) [inline WcaOnRails/app/models/competition.rb]: It's the label, I was trying to make it turn red...didn't work.
I'll change back to eventSpecs

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb (hunk 4 above)
Comment (pedrosino) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Done, done and done.

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb (hunk 5 above)
Comment (pedrosino) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Changed alert and moved css to competitions.css

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb (hunk 7 above)
Comment (pedrosino) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Done.

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb (hunk 6 above)
Comment (pedrosino) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Copied from the example :P 
Removed...

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb (hunk 8 above)
Comment (pedrosino) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Fixed.

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb (hunk 9 above)
Comment (pedrosino) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Done.

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb (hunk 10 above)
Comment (pedrosino) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Not necessary. It's gone now.

File: WcaOnRails/app/views/competitions/_nearby_competitions.html.erb [hunk 11]
```diff
@@ -54,8 +54,36 @@
       Last <%= Competition::NEARBY_DAYS_INFO %> days
     </label>
     <div class="col-sm-9">
-      <% nearby_count = @competition.nearby_competitions(Competition::NEARBY_DAYS_INFO, Competition::NEARBY_DISTANCE_KM_DANGER).count %>
-      <div class="<%= nearby_count < Competition::NEARBY_INFO_COUNT ? "alert alert-success" : "alert alert-danger" %>" role="alert"><%= nearby_count %> <%= "competition".pluralize(nearby_count) %> within <%= Competition::NEARBY_DISTANCE_KM_DANGER %>km and <%= Competition::NEARBY_DAYS_INFO %> days.
+      <% nearby_competitions_info = @competition.nearby_competitions(Competition::NEARBY_DAYS_INFO, Competition::NEARBY_DISTANCE_KM_INFO).sort_by(&:start_date) %>
+      <% nearby_count = nearby_competitions_info.count %>
+      <div class="alert <%= nearby_count < Competition::NEARBY_INFO_COUNT ? "alert-success" : "alert-danger" %>" role="alert"><%= nearby_count %> <%= "competition".pluralize(nearby_count) %> within <%= Competition::NEARBY_DISTANCE_KM_INFO %>km and <%= Competition::NEARBY_DAYS_INFO %> days.
+        <a id="show-modal" data-toggle="modal" data-target=".bs-example-modal-sm">Show</a>
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: This is an `<a>` tag, shouldn't it get cursor pointer without any css or an id?

Comment (timhabermaas) [inline WcaOnRails/app/views/competitions/_nearby_competitions.html.erb]: Not without an `href` AFAIK.

**Comment by jfly** [comment]
This has been merged and pushed live!

**Pull request closed without merging**